crates/re_data_source/src/load_file.rs

```diff
@@ -1,134 +1,255 @@
+                        store_id,
+                        path.clone(),
+                        Cow::Borrowed(&contents),
+                        tx_loader,
+                    ) {
+                        re_log::error!(?path, loader = loader.name(), %err, "failed to load data from file");
+                    }
+                } else {
+                    #[cfg(not(target_arch = "wasm32"))]
+                    if let Err(err) = loader.load_from_file(store_id, path.clone(), tx_loader) {
+                        re_log::error!(?path, loader = loader.name(), %err, "failed to load data from files");
+                    }
+                }
+            });
         }
 
-        // Assume an image (there are so many image formats)
-        _ => {
-            let format = if let Some(format) = ImageFormat::from_extension(extension) {
-                format
-            } else {
-                guess_format(&bytes).map_err(re_types::tensor_data::TensorImageLoadError::from)?
-            };
-
-            // Assume an image (there are so many image extensions):
-            let tensor = re_types::components::TensorData(
-                re_types::datatypes::TensorData::from_image_bytes(bytes, format)?,
-            );
-            Ok(vec![
-                image_indicator_cell(),
-                DataCell::try_from_native(std::iter::once(&tensor))?,
-            ])
+        // Implicitly closing `tx_loader`!
+
+        rx_loader
+    };
+
+    Ok(rx_loader)
+}
+
+/// Forwards the data in `rx_loader` to `tx`, taking care of necessary conversions, if any.
+///
+/// Runs asynchronously from another thread on native, synchronously on wasm.
+fn send(
+    store_id: &re_log_types::StoreId,
+    rx_loader: std::sync::mpsc::Receiver<LoadedData>,
+    tx: &Sender<LogMsg>,
+) {
+    spawn({
+        let tx = tx.clone();
+        let store_id = store_id.clone();
+        move || {
+            // ## Ignoring channel errors
+            //
+            // Not our problem whether or not the other end has hanged up, but we still want to
+            // poll the channel in any case so as to make sure that the data producer
+            // doesn't get stuck.
+            for data in rx_loader {
+                match data {
+                    LoadedData::DataRow(row) => {
+                        let mut table =
+                            re_log_types::DataTable::from_rows(re_log_types::TableId::new(), [row]);
+                        table.compute_all_size_bytes();
+
+                        let arrow_msg = table.to_arrow_msg().unwrap(); // TODO
+
+                        tx.send(LogMsg::ArrowMsg(store_id.clone(), arrow_msg)).ok();
+                    }
+
+                    LoadedData::ArrowMsg(msg) => {
+                        tx.send(LogMsg::ArrowMsg(store_id.clone(), msg)).ok();
+                    }
+
+                    LoadedData::LogMsg(msg) => {
+                        tx.send(msg).ok();
+                    }
+                }
```

Comment:
This could be a helper function on `LoadedData`:

```rs
impl LoadedData {
    pub fn into_log_msg(self, store_id: StoreId) -> LogMsg { … }
}
```

it would make this function shorter and more readable, and could be a useful function to have elsewhere